### ISSUE TITLE
When BuildLoop hits an unrecoverable failure, lorri daemon should exit

### ISSUE THREAD

**Issue by chreekat** [target/lorri#120]
Current behavior:

An unrecoverable failure (e.g. #66) kills the thread running the loop, but the lorri daemon keeps on truckin'. But it doesn't do anything. This means that a daemonized daemon must be babysat and restarted manually when hitting some unresolved "file not found" error over and over again.

Expected behavior:

If the looper thread perishes, the daemon should exit (or possibly restart the thread, but I vote for just dying and letting the system handle it).

**Comment by chreekat** [comment]
I should mention, when I said "e.g. #66", I mean my daemon is erroring out at the exact same call to unwrap(), as pasted below. To be clear, I don't give a hoot about the particular `Err` value, just the way the daemon handles it.


```
Jul 30 14:03:35 bryan-laptop lorri[7914]: thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Unrecoverable(Notify(Io(Os { code: 2, kind: NotFound, message: "No such file or directory" })))', src/libcore/result.rs:997:5
Jul 30 14:03:35 bryan-laptop lorri[7914]: stack backtrace:
Jul 30 14:03:35 bryan-laptop lorri[7914]:    0:     0x560ae72926df - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::hfad0220813d5e5c3
Jul 30 14:03:35 bryan-laptop lorri[7914]:    1:     0x560ae72a2d67 - std::sys_common::backtrace::print::h16017ddc8b354e39
Jul 30 14:03:35 bryan-laptop lorri[7914]:    2:     0x560ae72960a3 - std::panicking::default_hook::{{closure}}::h8b06d3b8a7138cc5
Jul 30 14:03:35 bryan-laptop lorri[7914]:    3:     0x560ae7295e23 - std::panicking::default_hook::h8cac4830452b35df
Jul 30 14:03:35 bryan-laptop lorri[7914]:    4:     0x560ae729679f - std::panicking::rust_panic_with_hook::hd32c26e1600a179b
Jul 30 14:03:35 bryan-laptop lorri[7914]:    5:     0x560ae7296331 - std::panicking::continue_panic_fmt::h24b78f7279f44cef
Jul 30 14:03:35 bryan-laptop lorri[7914]:    6:     0x560ae7296215 - rust_begin_unwind
Jul 30 14:03:35 bryan-laptop lorri[7914]:    7:     0x560ae72c3f7c - core::panicking::panic_fmt::h4cb3d77f52ba3d33
Jul 30 14:03:35 bryan-laptop lorri[7914]:    8:     0x560ae7168dbd - core::result::unwrap_failed::h7d5fda8dc4d9b8dc
Jul 30 14:03:35 bryan-laptop lorri[7914]:    9:     0x560ae715357a - lorri::build_loop::BuildLoop::forever::h939c8132ba23c652
Jul 30 14:03:35 bryan-laptop lorri[7914]:   10:     0x560ae714d879 - std::sys_common::backtrace::__rust_begin_short_backtrace::h6075ad49e64e4cb3
Jul 30 14:03:35 bryan-laptop lorri[7914]:   11:     0x560ae714e178 - std::panicking::try::do_call::h9294950418cd0877
Jul 30 14:03:35 bryan-laptop lorri[7914]:   12:     0x560ae72ada19 - __rust_maybe_catch_panic
Jul 30 14:03:35 bryan-laptop lorri[7914]:   13:     0x560ae7146517 - <F as alloc::boxed::FnBox<A>>::call_box::h4df23d87182ada14
Jul 30 14:03:35 bryan-laptop lorri[7914]:   14:     0x560ae72914ad - std::sys_common::thread::start_thread::h09a6bf95952964d1
Jul 30 14:03:35 bryan-laptop lorri[7914]:   15:     0x560ae729c105 - std::sys::unix::thread::Thread::new::thread_start::h165ad4e27a9c9315
Jul 30 14:03:35 bryan-laptop lorri[7914]:   16:     0x7f0bf8a22ef6 - start_thread
Jul 30 14:03:35 bryan-laptop lorri[7914]:   17:     0x7f0bf874222e - __GI___clone
Jul 30 14:03:35 bryan-laptop lorri[7914]:   18:                0x0 - <unknown>
```

**Comment by chreekat** [comment]
[Side note: I discovered what caused the unrecoverable error. My shell.nix was referencing an undefined file. With the same files, `nix-shell` reported:

```
$ nix-shell
error: getting status of '/home/b/RELEX/repos/the-repo/local-dev-environment/_generated-nix/google-oauth2.nix': No such file or directory
(use '--show-trace' to show detailed location information)
```

]

**Comment by Profpatsch** [comment]
> If the looper thread perishes, the daemon should exit (or possibly restart the thread, but I vote for just dying and letting the system handle it).

That sounds about right, we do want to kill the daemon in that case.

I’m gonna move the logic into this direction; is also means we need to be more careful in distinguishing between recoverable and unrecoverable errors (e.g. a missing nix file shouldn’t be an unrecoverable error methinks).

**Comment by grahamc** [comment]
We should avoid panicking in the build loop thread, but indeed, if it does panic the daemon should crash. We're probably not handling a `.join()` on the build loop thread properly.

**Comment by grahamc** [comment]
Closed by #156
